core/src/overlay.rs

```diff
@@ -0,0 +1,423 @@
+use std::mem::replace;
+use std::time::Instant;
+use std::{cell::RefCell, rc::Rc};
+
+use ribir_geom::{Size, ZERO_SIZE};
+use ribir_macros::Query;
+
+use crate::prelude::*;
+use crate::{
+  context::{AppCtx, BuildCtx, LayoutCtx},
+  widget::{BoxClamp, Render, WidgetBuilder, WidgetId},
+};
+
+#[derive(Clone)]
+pub struct OverlayStyle {
+  pub close_policy: ClosePolicy,
+  pub mask_brush: Option<Brush>,
+}
+
+bitflags! {
+  #[derive(Clone, Copy)]
+  pub struct ClosePolicy: u8 {
+    const NONE = 0b0000;
+    const ESC = 0b0001;
+    const TAP_OUTSIDE = 0b0010;
+  }
+}
+
+impl CustomStyle for OverlayStyle {
+  fn default_style(_: &BuildCtx) -> Self {
+    Self {
+      close_policy: ClosePolicy::ESC | ClosePolicy::TAP_OUTSIDE,
+      mask_brush: Some(Color::from_f32_rgba(0.3, 0.3, 0.3, 0.3).into()),
+    }
+  }
+}
+
+/// A handle to close the overlay
+#[derive(Clone)]
+pub struct OverlayCloseHandle(OverlayState);
+impl OverlayCloseHandle {
+  pub fn remove(&self) { self.0.remove() }
+}
+
+/// Use to control the overlay
+pub trait OverlayControl: OverlayBuilder {
```

Comment:
I think `OverlayControl` is unnecessary, directly implement methods for `Overlay` is better.